core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java

```diff
@@ -222,325 +262,407 @@ synchronized void updateGlobalCheckpointOnReplica(final long globalCheckpoint) {
             this.globalCheckpoint = globalCheckpoint;
             logger.trace("global checkpoint updated from primary to [{}]", globalCheckpoint);
         }
+        assert invariant();
     }
 
     /**
-     * Notifies the service of the current allocation ids in the cluster state. This method trims any shards that have been removed.
+     * Initializes the global checkpoint tracker in primary mode (see {@link #primaryMode}. Called on primary activation or promotion.
+     */
+    public synchronized void initializeAsPrimary(final String allocationId, final long localCheckpoint) {
```

Comment:
how would you feel about naming this method (and it's counterpart) `activatePrimaryMode` ? I was confused a couple of times as initialize and primary terms already used in the IndexShard context (a primary relocation target is a primary shard and is already initializing long before the method is called) .